---
title: "R Notebook"
output: html_notebook
---

## 0. Install packages
For the user who first running it, install required packages with:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
#BiocManager::install("biomaRt")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")
library(GEOmetadb)
if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")
```

## 1. Choose and download the data set
I chose my dataset on the website https://www.ncbi.nlm.nih.gov/geo/.

My dataset is GSE75168. The experiment samples the RNA sequence of cell lines MCF10A, MCF7, and MDA-MB-231 with 3 replicates. Genes in data with low expression (<3 counts) were removed from the analysis.

See more description on https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75168. 

### Information about platform

```{r, eval=FALSE}
# Get the GEO description of the dataset
gse <- getGEO("GSE75168",GSEMatrix=FALSE)

# Get information about the platform
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))
current_gpl_info$title
#[1] "Illumina HiSeq 1500 (Homo sapiens)"
current_gpl_info$last_update_date
#[1] "Oct 02 2018"
current_gpl_info$organism
#[1] "Homo sapiens"
length(current_gpl_info$series_id)
# 177
length(current_gpl_info$sample_id)
# 4012
```
As a summary:

_Experiment title__: 	Histone H3 lysine 4 acetylation-methylation dynamics define breast cancer subtypes [RNA-seq]

__Experiemtn type__: Expression profiling by high throughput sequencing

__Platform title__: Illumina NextSeq 500 (Homo sapiens)

__Submission data__: Feb 10, 2016

__LAst update data__: Mar 26 2019

__Organisms__: Homo sapiens

__Number of GEO datasets that use this techology__: 177

__Number of GEO samples that use this technology__: 4012

## 2. Data overview
Data can be download with:
```{r}
sfiles = getGEOSuppFiles('GSE75168')
fnames = rownames(sfiles)
ct_exp = read.delim(fnames[1],header=TRUE, check.names = FALSE)
# check.names = FALSE filter gname's version
head(ct_exp)
ct_exp_ver <- ct_exp
```
```{r}
#The version on ensembl_ID is deleted 
ct_exp[,1] <- stringr::str_split_fixed(ct_exp$Ensembl_ID, "\\.", 2)[,1]

head(ct_exp)
```
This part of code will be full explained in Mapping section. Basicly this part of code delete all version number of each ensembl id.

The data dimension is:
```{r}
dim(ct_exp)
```
Since there are 20575 rows of expression data, so it should be in good coverage.

The first column is Ensembl_ID, whcih should be mapped later.

The next 9 columns are the raw counts of 9 samples. Samples are categorized by:
```{r}
# Define the groups
samples <- data.frame(lapply(colnames(ct_exp)[2:10], 
                             FUN=function(x){unlist(strsplit(x, split = "\\_"))[c(1,2)]}))
colnames(samples) <- colnames(ct_exp)[2:10]
rownames(samples) <- c("cell_line", "replicate")
samples <- data.frame(t(samples),
                      stringsAsFactors=FALSE)
samples
```

The experiemnt contols the numbe rof replicate and tests three different cell lines, which are MCF10A, MCF7, MDA-MB-231. These cell lines are all from breast tissue. However, MCF10A is in luminal ductal cells for fibrocystic disease, while MCF7 and MDA-MB-231 are derived from metastatic site for adenocarcinoma.

## 3. Data Cleaning

```{r}
# Get the summarized counts for each gene
summarized_gene_counts <- sort(table(ct_exp$Ensembl_ID),decreasing = TRUE)
head(summarized_gene_counts)
```

It shows that my data on each row is for each gene, so the expression values are unique, and there are not nultiple rows to mapt the same symbol.


Also note that, as mentioned in Section 1, Genes in data with low expression (<3 counts) were removed from the analysis. So there should not be any outliers which can be removed.

Before normalization, weakly expressed and not-aligned vlaues can be filtered:
```{r}
#translate out counts into counts per million using the edgeR package function cpm
cpms = edgeR::cpm(ct_exp[,2:10])
rownames(cpms) <- ct_exp[,1]
# get rid of low counts (threshold as 3)
keep = rowSums(cpms >1) >=3
ct_exp_filtered = ct_exp[keep,]
# Check the difference
nrow(ct_exp)-nrow(ct_exp_filtered)
# Current covergae
dim(ct_exp_filtered)

```
There were 6297 rows are filtered. The filtered new data has 14278 rows of data, which is still in good coverage.

## 4. Normalization
Since normalization is not mentioned in either paper or the data proccessing chunk, my data should process the normalization:
```{r}
# Create an edgeR container for data
# Use filtered counts and make sure it is not a data.frame but id a matrix
# Give the defined groups that will be compared -> cell type (DK vs KP)
filtered_ct_filter <- as.matrix(ct_exp_filtered[, 2:10])
rownames(filtered_ct_filter) <- ct_exp_filtered$Ensembl_ID

d_filter = edgeR::DGEList(counts = filtered_ct_filter, group = samples$cell_line)
# Calculate the normalization factor
d_filter = edgeR::calcNormFactors(d_filter)
# Get the normalized data
normalized_counts <- edgeR::cpm(d_filter)
normalized_density <- apply(log2(normalized_counts[,1:9]), 2,
                            density)
```

### Comparaion the distrbution before and after normalization by graphs
Use boxplot to show the distribution before and after:
```{r warning = FALSE}
par(fig=c(0,0.48,0,1), new=TRUE)
unnormalized <- log2(ct_exp_filtered[, 2:10])
boxplot(unnormalized, xlab = "Samples", ylab = "log2 CPM", 
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Unnormalized")
#draw the median on each box plot
abline(h = median(apply(unnormalized, 2, median)), col = "green", lwd = 0.6, lty = "dashed")

par(fig=c(0.52,1,0,1), new=TRUE)
normalized <- log2(edgeR::cpm(normalized_counts[, 1:9]))
boxplot(normalized, xlab = "Samples", ylab = "log2 CPM", 
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Normalized")
#draw the median on each box plot
abline(h = median(apply(normalized, 2, median)), col = "green", lwd = 0.6, lty = "dashed")
```
which shows the normalized data indeed are more approached to the median.


And the plot:
```{r warning=FALSE}
dataPlot <- log2(edgeR::cpm(ct_exp_filtered[,2:10]))

# The graph before normalization
density <- apply(log2(ct_exp_filtered[,2:10]), 2, density)
xlim <- 0; ylim <- 0
for (i in 1:length(density)) {
  xlim <- range(c(xlim, density[[i]]$x));
  ylim <- range(c(ylim, density[[i]]$y))
}
cols <- rainbow(length(density))
ltys <- rep(1, length(density))

# Initialize the plot
plot(density[[1]], ylim=ylim, type="n",
     ylab="Unnormalized density of log2-CPM", main="", 
     cex.lab = 0.85, xlim = c(-10, 20))
# Plot each line
for (i in 1:length(density)) lines(density[[i]], col=cols[i], lty=ltys[i])
# Create legend
legend("topright",colnames(dataPlot),
       col=cols, lty=ltys, cex=0.55,
       border ="blue", text.col = "green4",
       merge = TRUE, bg = "gray90") 

# The plot after normalization
xlim <- 0; ylim <- 0
for (i in 1:length(normalized_density)) {
  xlim <- range(c(xlim, normalized_density[[i]]$x));
  ylim <- range(c(ylim, normalized_density[[i]]$y))
}
cols <- rainbow(length(normalized_density))
ltys <- rep(1, length(normalized_density))

# Initialize the plot
plot(normalized_density[[1]], xlim=c(-10, 20), ylim=ylim, type="n",
     ylab="Normalized density of log2-CPM", main="", cex.lab = 0.85)
# Plot each line
for (i in 1:length(normalized_density)) lines(normalized_density[[i]], col=cols[i], lty=ltys[i])
# Create legend
legend("topright",colnames(dataPlot),
       col=cols, lty=ltys, cex=0.55,
       border ="blue", text.col = "green4",
       merge = TRUE, bg = "gray90")
```
The plot comparaion also shows that there is a significant different distribution.

## 5. Mapping
Use biomaRt package to map ensembl id from column 1 in my data to HGNC symbol:
```{r eval=FALSE}
library(biomaRt)
mart <- biomaRt::useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "useast")

conversion_stash <- "_id_conversion.rds"
if(file.exists(conversion_stash)){
  id_conversion_ver <- readRDS(conversion_stash)
} else {
  id_conversion_ver <- getBM(attributes = c("ensembl_gene_id_version","hgnc_symbol"),
                         filters = c("ensembl_gene_id_version"),
                         values = ct_exp_ver$Ensembl_ID,
                         mart = mart)
  saveRDS(id_conversion_ver, conversion_stash)
}

```
The attributes and filters both use __ensembl_gene_id_version__ since there is version number for each ensembl id of my data (see Section 2).

```{r}
nrow(ct_exp) - nrow(id_conversion_ver)
nrow(ct_exp)
```
However, it shows that most of gene from the raw data cannot be mapped, which is not good. After I went over my whole data, I found that the poor mappping problem may be because of the version number of 

So this is why all version numbers were deleted in the Data Overview section.

So let's run the mapping again:

```{r eval = FALSE}
conversion_stash <- "id_conversion.rds"
if(file.exists(conversion_stash)){
  id_conversion <- readRDS(conversion_stash)
} else {
  id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                         filters = c("ensembl_gene_id"),
                         values = ct_exp_filtered$Ensembl_ID,
                         mart = mart)
  saveRDS(id_conversion, conversion_stash)
}
```

And the number of missing identifers is:
```{r}
nrow(ct_exp_filtered) - nrow(id_conversion)
nrow(ct_exp_filtered)
```
which is much better than before.

And the final coverage becomes
```{r}
14278-91
```
which is still a good coverage.

## 6. Sample seperation
Use multidimentional scalling plot:
```{r}
plotMDS(d_filter, labels=rownames(samples),
        col = c("darkgreen","blue", "red")[factor(samples$cell_line)], cex=0.5)

```

So samples in each test group are in small distance within their group by compare to the test and control group. 

This shows that samples in test group are more silimar to each other rather than to the data of control group. 

## Reference
Messier TL, Gordon JA, Boyd JR, Tye CE et al. Histone H3 lysine 4 acetylation and methylation dynamics define breast cancer subtypes. Oncotarget 2016 Feb 2;7(5):5094-109. PMID: 26783963

